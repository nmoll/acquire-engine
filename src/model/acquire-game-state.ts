export type AcquireGameState = "not started" | "started";
