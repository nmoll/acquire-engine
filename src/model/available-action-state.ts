import { AvailableAction } from "./available-action";

export type IAvailableActionState = AvailableAction[];
