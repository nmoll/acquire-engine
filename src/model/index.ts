export * from "./board-square-state";
export * from "./game-state";
export * from "./hotel-chain-type";
export * from "./player-turn";
export * from "./shares-state";
