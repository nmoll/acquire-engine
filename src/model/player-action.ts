import { HotelChainType } from "./hotel-chain-type";
import { IShares } from "./shares";

export interface PlaceTile {
  type: "PlaceTile";
  playerId: number;
  boardSquareId: number;
}

export interface StartHotelChain {
  type: "StartHotelChain";
  playerId: number;
  hotelChain: HotelChainType;
}

export interface PurchaseShares {
  type: "PurchaseShares";
  playerId: number;
  shares: IShares[];
}

export interface ChooseMergeDirection {
  type: "ChooseMergeDirection";
  playerId: number;
  hotelChainToKeep: HotelChainType;
}

export type PlayerAction =
  | PlaceTile
  | StartHotelChain
  | PurchaseShares
  | ChooseMergeDirection;

export const PlayerActionType = {
  PlaceTile: (playerId: number, boardSquareId: number): PlaceTile => ({
    type: "PlaceTile",
    playerId,
    boardSquareId,
  }),
  StartHotelChain: (
    playerId: number,
    hotelChain: HotelChainType
  ): StartHotelChain => ({
    type: "StartHotelChain",
    playerId,
    hotelChain,
  }),

  PurchaseShares: (playerId: number, shares: IShares[]): PurchaseShares => ({
    type: "PurchaseShares",
    playerId,
    shares,
  }),

  ChooseMergeDirection: (
    playerId: number,
    hotelChainToKeep: HotelChainType
  ): ChooseMergeDirection => ({
    type: "ChooseMergeDirection",
    playerId,
    hotelChainToKeep,
  }),
};
