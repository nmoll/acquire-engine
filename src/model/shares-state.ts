export interface ISharesState {
  [playerId: string]: {
    ["American"]?: number;
    ["Continental"]?: number;
    ["Festival"]?: number;
    ["Imperial"]?: number;
    ["Luxor"]?: number;
    ["Tower"]?: number;
    ["Worldwide"]?: number;
  };
}
