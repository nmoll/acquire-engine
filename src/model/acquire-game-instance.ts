import { AcquireGameState } from "./acquire-game-state";

export interface IAcquireGameInstance {
  id: string;
  randomSeed: number;
  playerIds: string[];
  hostId: string;
  state: AcquireGameState;
  isOpen?: boolean;
  createdDate?: number;
}
