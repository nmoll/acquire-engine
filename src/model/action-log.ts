import { IGameState } from "./game-state";
import { PlayerAction } from "./player-action";
import { PlayerActionResult } from "./player-action-result";

export interface ActionLog {
  action: PlayerAction;
  actionResult: PlayerActionResult;
  state: IGameState;
}
