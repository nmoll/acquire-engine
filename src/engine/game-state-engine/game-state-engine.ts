import { IGameState } from "../../model";
import { IAcquireGameInstance } from "../../model/acquire-game-instance";
import { ActionLog } from "../../model/action-log";
import { PlayerAction } from "../../model/player-action";
import { ActionResultEngine } from "../action-result-engine/action-result-engine";
import { AvailableActionsStateEngine } from "../available-actions-state-engine/available-actions-state-engine";
import { BoardStateEngine } from "../board-state-engine/board-state-engine";
import { CashStateEngine } from "../cash-state-engine/cash-state-engine";
import { CurrentPlayerIdStateEngine } from "../current-player-id-state-engine/current-player-id-state-engine";
import { SharesStateEngine } from "../shares-state-engine/shares-state-engine";
import { TileStateEngine } from "../tile-state-engine/tile-state-engine";

const getInitialState = (gameInstance: IAcquireGameInstance): IGameState => {
  const boardState = BoardStateEngine.computeState();
  const cashState = CashStateEngine.computeState(gameInstance);
  const tileState = TileStateEngine.computeState(gameInstance);
  const sharesState = SharesStateEngine.computeState(gameInstance);
  const currentPlayerIdState =
    CurrentPlayerIdStateEngine.computeState(gameInstance);
  const availableActionsState = AvailableActionsStateEngine.computeState(
    boardState,
    sharesState,
    cashState
  );

  return {
    boardState,
    cashState,
    tileState,
    sharesState,
    currentPlayerIdState,
    availableActionsState,
  };
};

const computeGameState = (
  gameInstance: IAcquireGameInstance,
  playerActions: PlayerAction[]
): IGameState => {
  const gameLog: ActionLog[] = [];

  return playerActions.reduce<IGameState>((state, action) => {
    if (!AvailableActionsStateEngine.validateAction(action, state)) {
      return state;
    }

    const actionResult = ActionResultEngine.computeActionResult(state, action);

    const boardState = BoardStateEngine.computeState(
      actionResult,
      state.boardState
    );

    const cashState = CashStateEngine.computeState(
      gameInstance,
      state,
      actionResult,
      gameLog
    );

    const tileState = TileStateEngine.computeState(
      gameInstance,
      actionResult,
      state.tileState
    );

    const sharesState = SharesStateEngine.computeState(
      gameInstance,
      actionResult,
      state.sharesState
    );

    const currentPlayerIdState = CurrentPlayerIdStateEngine.computeState(
      gameInstance,
      actionResult,
      sharesState,
      gameLog
    );

    const availableActionsState = AvailableActionsStateEngine.computeState(
      boardState,
      sharesState,
      cashState,
      actionResult,
      gameLog
    );

    const newState = {
      boardState,
      cashState,
      sharesState,
      tileState,
      availableActionsState,
      currentPlayerIdState,
    };

    gameLog.push({
      action,
      actionResult,
      state: newState,
    });

    return newState;
  }, getInitialState(gameInstance));
};

export const GameStateEngine = {
  computeGameState,
};
