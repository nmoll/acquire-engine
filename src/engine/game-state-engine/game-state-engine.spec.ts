import { BoardStateFactory } from "../../../test/factory/board-state.factory";
import { SharesStateFactory } from "../../../test/factory/shares-state.factory";
import { america, getTilePosition, imperial } from "../../../test/helpers";
import { HotelChainType, IGameState } from "../../model";
import { PlayerAction, PlayerActionType } from "../../model/player-action";
import { GameStateEngine } from "./game-state-engine";

describe("GameStateEngine", () => {
  it("should compute initial game state with no actions", () => {
    const playerIds = [1, 2, 3, 4];
    const actions: PlayerAction[] = [];

    const expectedState: IGameState = {
      boardState: BoardStateFactory.createBoardState(
        `
          - - - - - - - - - - - -
          - - - - - - - - - - - -
          - - - - - - - - - - - -
          - - - - - - - - - - - -
          - - - - - - - - - - - -
          - - - - - - - - - - - -
          - - - - - - - - - - - -
          - - - - - - - - - - - -
          - - - - - - - - - - - -`
      ),
      cashState: {},
      sharesState: {},
      currentPlayerIdState: null,
    };

    expect(GameStateEngine.computeGameState(playerIds, actions)).toEqual(
      expectedState
    );
  });

  it("should compute game state with multiple actions", () => {
    const playerIds = [1, 2, 3, 4];
    const actions: PlayerAction[] = [
      PlayerActionType.PlaceTile(1, getTilePosition("5A")),

      PlayerActionType.PlaceTile(2, getTilePosition("2D")),

      PlayerActionType.PlaceTile(3, getTilePosition("3D")),
      PlayerActionType.StartHotelChain(3, HotelChainType.AMERICAN),
      PlayerActionType.PurchaseShares(3, [america(3)]),

      PlayerActionType.PlaceTile(4, getTilePosition("4F")),
      PlayerActionType.PurchaseShares(4, [america(2)]),

      PlayerActionType.PlaceTile(1, getTilePosition("10D")),

      PlayerActionType.PlaceTile(2, getTilePosition("5G")),
      PlayerActionType.PurchaseShares(2, [america(3)]),

      PlayerActionType.PlaceTile(3, getTilePosition("5F")),
      PlayerActionType.StartHotelChain(3, HotelChainType.IMPERIAL),
      PlayerActionType.PurchaseShares(3, [america(1), imperial(2)]),

      PlayerActionType.PlaceTile(4, getTilePosition("3F")),
      PlayerActionType.PurchaseShares(4, [america(3)]),

      PlayerActionType.PlaceTile(1, getTilePosition("3E")),
    ];

    const expectedState: IGameState = {
      boardState: BoardStateFactory.createBoardState(` 
          - - - - 0 - - - - - - -
          - - - - - - - - - - - -
          - - - - - - - - - - - -
          - I I - - - - - - 0 - -
          - - I - - - - - - - - -
          - - I I I - - - - - - -
          - - - - I - - - - - - - 
          - - - - - - - - - - - -
          - - - - - - - - - - - -`),
      cashState: {
        1: 6000,
        2: 5100,
        3: 4200,
        4: 4500,
      },
      sharesState: SharesStateFactory.createSharesState(`
             A C F I L T W
          P1 0 0 0 0 0 0 0
          P2 3 0 0 0 0 0 0
          P3 5 0 0 3 0 0 0
          P4 5 0 0 0 0 0 0
          `),
      currentPlayerIdState: 1,
    };

    expect(GameStateEngine.computeGameState(playerIds, actions)).toEqual(
      expectedState
    );
  });
});
