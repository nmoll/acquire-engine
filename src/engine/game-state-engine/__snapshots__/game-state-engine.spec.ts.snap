// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameStateEngine should create tile map 1`] = `
Object {
  "0": "1A",
  "1": "2A",
  "10": "11A",
  "100": "5I",
  "101": "6I",
  "102": "7I",
  "103": "8I",
  "104": "9I",
  "105": "10I",
  "106": "11I",
  "107": "12I",
  "11": "12A",
  "12": "1B",
  "13": "2B",
  "14": "3B",
  "15": "4B",
  "16": "5B",
  "17": "6B",
  "18": "7B",
  "19": "8B",
  "2": "3A",
  "20": "9B",
  "21": "10B",
  "22": "11B",
  "23": "12B",
  "24": "1C",
  "25": "2C",
  "26": "3C",
  "27": "4C",
  "28": "5C",
  "29": "6C",
  "3": "4A",
  "30": "7C",
  "31": "8C",
  "32": "9C",
  "33": "10C",
  "34": "11C",
  "35": "12C",
  "36": "1D",
  "37": "2D",
  "38": "3D",
  "39": "4D",
  "4": "5A",
  "40": "5D",
  "41": "6D",
  "42": "7D",
  "43": "8D",
  "44": "9D",
  "45": "10D",
  "46": "11D",
  "47": "12D",
  "48": "1E",
  "49": "2E",
  "5": "6A",
  "50": "3E",
  "51": "4E",
  "52": "5E",
  "53": "6E",
  "54": "7E",
  "55": "8E",
  "56": "9E",
  "57": "10E",
  "58": "11E",
  "59": "12E",
  "6": "7A",
  "60": "1F",
  "61": "2F",
  "62": "3F",
  "63": "4F",
  "64": "5F",
  "65": "6F",
  "66": "7F",
  "67": "8F",
  "68": "9F",
  "69": "10F",
  "7": "8A",
  "70": "11F",
  "71": "12F",
  "72": "1G",
  "73": "2G",
  "74": "3G",
  "75": "4G",
  "76": "5G",
  "77": "6G",
  "78": "7G",
  "79": "8G",
  "8": "9A",
  "80": "9G",
  "81": "10G",
  "82": "11G",
  "83": "12G",
  "84": "1H",
  "85": "2H",
  "86": "3H",
  "87": "4H",
  "88": "5H",
  "89": "6H",
  "9": "10A",
  "90": "7H",
  "91": "8H",
  "92": "9H",
  "93": "10H",
  "94": "11H",
  "95": "12H",
  "96": "1I",
  "97": "2I",
  "98": "3I",
  "99": "4I",
}
`;
